R/class-Collections.R

```diff
@@ -0,0 +1,415 @@
+            msg <- paste("All columns must belong to the same entity.")
+            errors <- c(errors, msg)
+        }
+    }
+
+    if (length(errors) == 0) {
+        return(TRUE)
+    } else {
+        return(errors)
+    }
+}
+
+#' Data Collection
+#' 
+#' This class represents a collection of related data. That could be
+#' all of the abundance data for all samples in a dataset at a particular taxonomic rank,
+#' or all pathway abundance data for all samples in a dataset, or something else. The
+#' primary requirement for a collection is that the values of all variables in the
+#' collection can be expressed on the same theoretical range.
+#' @slot name The name of the collection
+#' @slot data A data.frame of data with samples as rows and variables as columns
+#' @slot recordIdColumn The name of the column in the data.frame that contains the record id
+#' @slot ancestorIdColumns A character vector of column names representing parent entities of the recordIdColumn
+#' @slot imputeZero A logical indicating whether NA/ null values should be replaced with zeros.
+#' @slot removeEmptyRecordss A logical indicating whether empty (all NA/ zero) records should be removed.
+#' @name Collection-class
+#' @rdname Collection-class
+#' @export
+setClass("Collection", 
+    slots = c(
+        name = "character",
+        data = "data.table",
+        recordIdColumn = "character",
+        ancestorIdColumns = "character",
+        imputeZero = 'logical',
+        removeEmptyRecords = 'logical'
+    ), prototype = prototype(
+        imputeZero = TRUE,
+        removeEmptyRecords = TRUE
+    ),
+    validity = check_collection
+)
+
+
+check_collections <- function(object) {
+    errors <- character()
+
+    # check that all names are unique
+    if (length(unique(getCollectionNames(object))) != length(object)) {
+        msg <- "collection names must be unique"
+        errors <- c(errors, msg)
+    }
+
+    if (length(object) == 0) {
+        return(TRUE)
+    }
+
+    # check that at least one ancestorIdColumn is shared between collections
+    firstCollectionAncestorIds <- object[[1]]@ancestorIdColumns
+    if (!all(sapply(object, function(x) any(x@ancestorIdColumns %in% firstCollectionAncestorIds)))) {
+        msg <- "at least one ancestorIdColumn must be shared between collections"
+        errors <- c(errors, msg)
+    }
+
+    if (length(errors) == 0) {
+        return(TRUE)
+    } else {
+        return(errors)
+    }
+}
+
+#' Data Collections
+#' 
+#' This is a list of Data Collections.
+#' @name Collections-class
+#' @rdname Collections-class
+#' @importFrom S4Vectors SimpleList
+setClass("Collections", 
+    contains = "SimpleList",
+    prototype = prototype(
+        elementType = "Collection"
+    ),
+    validity = check_collections
+)
+
+#' Create a Collection
+#' 
+#' This is a constructor for the Collection class. It creates a Collection
+#' object for use in a Microbiome Dataset. 
+#' @param name The name of the collection
```

Comment:
Can we add examples here? What kind of name might be useful? Or when would i use the name later?

Reply:
this is for anyone who interacts w these data via the R cli.. so they know what the collections is, conceptually. so the name can be anything that aids understanding. i can try to clarify the documentation.